### PR TITLE
Pipelines CLI Fixes

### PR DESCRIPTION
**Pull request by null-sleep**
This PR makes the following changes to the Pipelines CLI:

- A successful spec deployment prints the URL to the pipeline.
- If the provided spec file does not have an ID, one is generated and the spec file is updated with the new ID.
- Adds a new exception eater to print cleaner error messages from the pipelines API
- Adds exception handing for specs with invalid json
- A validation check for pipeline id: only -, _ and alphanumeric characters are allowed

Testing:
Unit Tests
Manually verified deployment URLs, new exception formatting, validation of pipeline ids

### CODE REVIEW

File: databricks_cli/utils.py [hunk 1]
```diff
@@ -46,9 +46,39 @@ def decorator(*args, **kwargs):
                                'reconfigure with ``dbfs configure``')
             else:
                 error_and_quit(exception.response.content)
-        except Exception as exception: # noqa
+        except Exception as exception:  # noqa
             if not DEBUG_MODE:
                 error_and_quit('{}: {}'.format(type(exception).__name__, str(exception)))
+
+    decorator.__doc__ = function.__doc__
```
Comment (null-sleep) [inline databricks_cli/utils.py:53]: This is just formatting

File: databricks_cli/pipelines/cli.py [hunk 2]
```diff
@@ -59,8 +69,20 @@ def deploy_cli(api_client, spec_arg, spec):
         raise RuntimeError('The spec should be provided either by an option or argument')
     src = spec_arg if bool(spec_arg) else spec
     spec_obj = _read_spec(src)
+    if 'id' not in spec_obj:
+        pipeline_id = str(uuid.uuid4())
+        click.echo("Updating spec with id: {}".format(pipeline_id))
+        spec_obj['id'] = pipeline_id
+        _write_spec(src, spec_obj)
+    elif not _validate_pipeline_id(spec_obj['id']):
+        error_and_quit("Pipeline id has invalid characters")
```
Comment (mukulmurthy) [inline databricks_cli/pipelines/cli.py:78]: Say what's allowed and also print the id they provided. Also, maybe the error_and_quit should just be inside the _validate_pipeline_id call so you don't have to repeat those bits?

File: databricks_cli/pipelines/cli.py [hunk 3]
```diff
@@ -59,8 +69,20 @@ def deploy_cli(api_client, spec_arg, spec):
...
     PipelinesApi(api_client).deploy(spec_obj)
 
+    pipeline_id = spec_obj['id']
+    base_url = "{0.scheme}://{0.netloc}/".format(urlparse(api_client.url))
+    pipeline_url = urljoin(base_url, "#joblist/pipelines/{}".format(pipeline_id))
```
Comment (mukulmurthy) [inline databricks_cli/pipelines/cli.py:83]: File a task on our side to return this from deployment (and then also update this to use that).

Comment (mukulmurthy) [reply databricks_cli/pipelines/cli.py:83]: Does this work for MT shards?

File: databricks_cli/pipelines/cli.py [hunk 4]
```diff
@@ -59,8 +69,20 @@ def deploy_cli(api_client, spec_arg, spec):
...
+    click.echo(pipeline_url)
```
Comment (mukulmurthy) [inline databricks_cli/pipelines/cli.py:84]: Better message, like "Pipeline {id} has been deployed and can be viewed at url"

File: databricks_cli/utils.py [hunk 5]
```diff
@@ -46,9 +46,39 @@ def decorator(*args, **kwargs):
...
+    """
+
+    @six.wraps(function)
+    def decorator(*args, **kwargs):
+        try:
+            return function(*args, **kwargs)
+        except HTTPError as exception:  # noqa
+            if exception.response.status_code == 401:
+                error_and_quit('Your authentication information may be incorrect. Please '
+                               + 'reconfigure with ``dbfs configure``')
```
Comment (mukulmurthy) [inline databricks_cli/utils.py:70]: databricks configure

File: databricks_cli/utils.py [hunk 6]
```diff
@@ -46,9 +46,39 @@ def decorator(*args, **kwargs):
...
+
+            elif exception.response.status_code == 400:
+                exp_context = json_loads(exception.response.content.decode('utf-8'))
+                message = exp_context['error_code'] + '\n' + exp_context['message']
```
Comment (mukulmurthy) [inline databricks_cli/utils.py:74]: Can we do this for all exceptions first (even before the 401), and only do the other ones if it fails to get this info?

File: tests/pipelines/test_cli.py [hunk 7]
```diff
@@ -124,6 +124,19 @@ def test_delete_cli_incorrect_parameters(pipelines_api_mock, tmpdir):
...
+def test_deploy_spec_updated_with_id_if_pipeline_id_not_in_spec(pipelines_api_mock, tmpdir):  # noqa
+    path = tmpdir.join('/spec.json').strpath
+    spec_with_no_id = '{"name": "no id"}'
+    with open(path, 'w') as f:
+        f.write(spec_with_no_id)
+    runner = CliRunner()
+    runner.invoke(cli.deploy_cli, ['--spec', path])
+    with open(path, 'r') as f:
+        updated_spec = json.loads(f.read())
+    assert 'id' in updated_spec
```
Comment (mukulmurthy) [inline tests/pipelines/test_cli.py:137]: Couple missing tests for not touching the ID if it's valid, and what happens if the id is an empty string.

Comment (mukulmurthy) [reply tests/pipelines/test_cli.py:137]: Let's also test for invalid json?

File: databricks_cli/pipelines/cli.py [hunk 8]
```diff
@@ -21,16 +21,26 @@
...
-from json import loads as json_loads
 import os
+import uuid
+import json
+import string
+
+try:
+    from urlparse import urlparse, urljoin
+except ImportError:
+    from urllib.parse import urlparse, urljoin
```
[the remark this answers is not in the document]
Comment (null-sleep) [reply databricks_cli/pipelines/cli.py:32]: It's a pattern used to support python2/3. The library was renamed in py3

File: databricks_cli/utils.py (hunk 5 above)
Comment (null-sleep) [reply databricks_cli/utils.py:70]: this part is replicating `eat_exceptions`. if we want to change this, it should be in a separate PR and we should do it in both places

File: databricks_cli/pipelines/cli.py (hunk 3 above)
Comment (null-sleep) [reply databricks_cli/pipelines/cli.py:83]: Yes `api_client.url` cleans up the `?o=123` part and just returns the host url

File: tests/pipelines/test_cli.py (hunk 7 above)
Comment (null-sleep) [reply tests/pipelines/test_cli.py:137]: Invalid json input is tested in `_read_spec`. I think validate should raise an error for an empty or invalid id and I'll add a test for the case when an id is provided.

File: databricks_cli/utils.py (hunk 6 above)
Comment (null-sleep) [reply databricks_cli/utils.py:74]: Have a handler that tries to parse for error_code and message and if it fails, raises the Exception?
